common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java

```diff
@@ -405,19 +469,35 @@ protected boolean runAllTasks(long timeoutNanos) {
             if ((runTasks & 0x3F) == 0) {
                 lastExecutionTime = ScheduledFutureTask.nanoTime();
                 if (lastExecutionTime >= deadline) {
+                    result = DeadlineRunResult.TIMED_OUT;
```

Comment:
> Yes it removes the enum of return type but introduces a complexity that ties a callback of 
execution time and return value

Currently the return value indicates `timed_out` however the actual timing information is communicated through the callback. It is not clear what additional value this return value provides and duplicating the responsibilities makes the interface more complex IMHO. For example what if return value says TIMED_OUT but the callback time indicates there is still time to spare? This may be a programming error but why have an interface where these ambiguities are possible if it is not necessary?

> I consider this callback as a side-effect that happens to impact SingleThreadEventLoop but not when called from SingleThreadEventExecutor.

I view this as a general side effect of the requirements of the interface contract. If multiple calls must be made and timing information must be preserved across these calls then. This currently manifests itself in `SingleThreadEventLoop` but keep in mind these methods are `protected` and the visibility is not limited to `SingleThreadEventLoop`.

the main differences I see are:

- Preserving the timeout semantics when running tailTasks from `SingleThreadEventLoop.runAllTasks(long timeoutNanos)`.
- Separation of concern between time consumed by the method execution, and "was a task run" (which already must exist to preserve timing across multiple calls to this method)
- Consolidates timeout logic into the callback instead of duplicating this in the return value and callback.